File: src/lib/services/tagCountService.ts
import { adminDb } from '@/lib/firebase/admin';

/**
 * Updates the contentCount for a specific tag based on content that uses it
 */
export async function updateTagCount(userId: string, tagId: string): Promise<void> {
  if (!tagId || !userId) return;

  try {
    // Count content items that have this tagId in their tagsId array
    const contentCollection = adminDb
      .collection('users')
      .doc(userId)
      .collection('content');
    
    const contentSnapshot = await contentCollection.where('tagsId', 'array-contains', tagId).get();
    const currentContentCount = contentSnapshot.size;

    // Update the tag's contentCount
    const tagRef = adminDb
      .collection('users')
      .doc(userId)
      .collection('tags')
      .doc(tagId);

    // Check if tag exists before updating
    const tagDoc = await tagRef.get();
    if (tagDoc.exists) {
      await tagRef.update({
        contentCount: currentContentCount,
        updatedAt: new Date().toISOString()
      });
    }
  } catch (error) {
    console.error('Error updating tag count:', error);
  }
}

/**
 * Updates tag counts for multiple tags
 */
export async function updateMultipleTagCounts(userId: string, tagIds: string[]): Promise<void> {
  const validTagIds = tagIds.filter(id => id && id.trim() !== '');
  
  // Update each tag count
  await Promise.all(
    validTagIds.map(tagId => updateTagCount(userId, tagId))
  );
}

/**
 * Updates tag counts when content changes from one tag to another
 */
export async function updateTagCountsOnContentChange(
  userId: string, 
  oldTagId?: string, 
  newTagId?: string
): Promise<void> {
  const tagsToUpdate: string[] = [];
  
  if (oldTagId && oldTagId.trim() !== '') {
    tagsToUpdate.push(oldTagId);
  }
  
  if (newTagId && newTagId.trim() !== '' && newTagId !== oldTagId) {
    tagsToUpdate.push(newTagId);
  }
  
  await updateMultipleTagCounts(userId, tagsToUpdate);
}

/**
 * Updates tag counts when content changes from old tag array to new tag array
 */
export async function updateTagCountsOnArrayChange(
  userId: string, 
  oldTagIds: string[] = [], 
  newTagIds: string[] = []
): Promise<void> {
  // Get all unique tag IDs that need count updates
  const allTagIds = new Set([...oldTagIds, ...newTagIds]);
  
  // Update counts for all affected tags
  await updateMultipleTagCounts(userId, Array.from(allTagIds));
}

/**
 * Recalculates tag counts for all tags of a user
 * Useful for syncing up counts if they get out of sync
 */
export async function recalculateAllTagCounts(userId: string): Promise<void> {
  try {
    // Get all tags for the user
    const tagsCollection = adminDb
      .collection('users')
      .doc(userId)
      .collection('tags');
    
    const tagsSnapshot = await tagsCollection.get();
    
    // Update count for each tag
    const updatePromises = tagsSnapshot.docs.map(async (tagDoc) => {
      const tagId = tagDoc.id;
      await updateTagCount(userId, tagId);
    });
    
    await Promise.all(updatePromises);
  } catch (error) {
    console.error('Error recalculating all tag counts:', error);
    throw error;
  }
}

File: backend/app/services/tag_count_service.py
import asyncio
from app.services.firebase.firebase_admin_setup import admin_db
from datetime import datetime
from typing import List, Optional


async def update_tag_count(user_id: str, tag_id: str) -> None:
    if not tag_id or not user_id: return

    try:
        content_collection = admin_db.collection('users').document(user_id).collection('content')
        
        content_snapshot = content_collection.where('tagsId', 'array_contains', tag_id).get()
        current_content_count = len(content_snapshot)

        tag_ref = admin_db.collection('users').document(user_id).collection('tags').document(tag_id)
        
        tag_doc = tag_ref.get()
        if tag_doc.exists:
            tag_ref.update({
                'contentCount': current_content_count,
                'updatedAt': datetime.utcnow().isoformat() + "Z"
            })
    except Exception as error:
        print(f"Error updating tag count: {error}")

async def update_multiple_tag_counts(user_id: str, tag_ids: list[str]) -> None:
    valid_tag_ids = [tid for tid in tag_ids if tid and tid.strip() != '']
    if valid_tag_ids:
        await asyncio.gather(*(update_tag_count(user_id, tid) for tid in valid_tag_ids))


async def update_tag_counts_on_content_change(
    user_id: str, 
    old_tag_id: Optional[str] = None, 
    new_tag_id: Optional[str] = None
) -> None:
    """
    Updates tag counts when content changes from one tag to another
    """
    tags_to_update: List[str] = []
    
    if old_tag_id and old_tag_id.strip() != '':
        tags_to_update.append(old_tag_id)
    
    if new_tag_id and new_tag_id.strip() != '' and new_tag_id != old_tag_id:
        tags_to_update.append(new_tag_id)
    
    await update_multiple_tag_counts(user_id, tags_to_update)


async def update_tag_counts_on_array_change(
    user_id: str, 
    old_tag_ids: List[str] = [], 
    new_tag_ids: List[str] = []
) -> None:
    """
    Updates tag counts when content changes from old tag array to new tag array
    """
    # Get all unique tag IDs that need count updates
    all_tag_ids = set(old_tag_ids + new_tag_ids)
    
    # Update counts for all affected tags
    await update_multiple_tag_counts(user_id, list(all_tag_ids))


async def recalculate_all_tag_counts(user_id: str) -> None:
    """
    Recalculates tag counts for all tags of a user
    Useful for syncing up counts if they get out of sync
    """
    try:
        # Get all tags for the user
        tags_collection = admin_db.collection('users').document(user_id).collection('tags')
        tags_snapshot = tags_collection.get()
        
        # Update count for each tag
        # Replicating tagsSnapshot.docs.map and Promise.all
        tasks = []
        for tag_doc in tags_snapshot:
            tag_id = tag_doc.id
            tasks.append(update_tag_count(user_id, tag_id))
        
        if tasks:
            await asyncio.gather(*tasks)
            
    except Exception as error:
        # Replicating console.error and throw error
        print(f"Error recalculating all tag counts: {error}")
        raise error